### PR TITLE
drivers: dp: swdp_bitbang: power optimization

### PR DESCRIPTION
**Pull request by maxd-nordic**
This patch changes GPIO initialization
to be in PORT_OFF state by default.
Also, if no transceiver is attached to the signals, the GPIOs are configured to be disconnected to preserve power.

I tested this on a prototype board where we are going to have a debugger in a low-power context.

### CODE REVIEW

**Comment by maxd-nordic** [comment]
@fabiobaltieri @henrikbrixandersen @mniestroj @tmon-nordic @jfischer-no Could you have a look?

**Review by fabiobaltieri — DISMISSED** [review]

**Comment by fabiobaltieri** [comment]
You folks probably wants to set a maintainers area for this.

**Review by mniestroj — DISMISSED** [review]

**Review by tmon-nordic — DISMISSED** [review]

File: drivers/dp/swdp_bitbang.c [hunk 1]
```diff
@@ -562,113 +562,120 @@ static int sw_configure(const struct device *dev,
...
+	int ret;
 
 	if (config->dnoe.port) {
-		gpio_pin_set_dt(&config->dnoe, 1);
+		ret = gpio_pin_configure_dt(&config->dnoe, GPIO_OUTPUT_ACTIVE);
+		if (ret) {
+			return ret;
+		}
 	}
-
```
Comment (jfischer-no) [inline drivers/dp/swdp_bitbang.c:571]: Stray change.

File: drivers/dp/swdp_bitbang.c [hunk 2]
```diff
@@ -562,113 +562,120 @@ static int sw_configure(const struct device *dev,
...
+	if (ret) {
+		return ret;
+	}
+
 	sw_data->turnaround = 1U;
 	sw_data->data_phase = false;
 	sw_data->fast_clock = false;
 	sw_data->clock_delay = CLOCK_DELAY(SWDP_DEFAULT_SWCLK_FREQUENCY,
 					   config->port_write_cycles);
-
```
Comment (jfischer-no) [inline drivers/dp/swdp_bitbang.c:671]: Stray change.

File: drivers/dp/swdp_bitbang.c [hunk 3]
```diff
@@ -562,113 +562,120 @@ static int sw_configure(const struct device *dev,
...
+	if (config->dnoe.port) {
+		ret = gpio_pin_configure_dt(&config->dnoe, GPIO_OUTPUT_INACTIVE);
 		if (ret) {
 			return ret;
 		}
-	}
-
-	if (config->dnoe.port) {
-		ret = gpio_pin_configure_dt(&config->dnoe, GPIO_OUTPUT_INACTIVE);
+		if (config->dout.port) {
```
Comment (jfischer-no) [inline drivers/dp/swdp_bitbang.c:612]:
```c
	if (config->dnoe.port) {
		ret = gpio_pin_configure_dt(&config->dnoe, GPIO_OUTPUT_INACTIVE);
		if (ret) {
			return ret;
		}

		if (config->dout.port) {
```
Please fix in all your changes accordingly.

**Review by jfischer-no — CHANGES_REQUESTED** [review]

File: drivers/dp/swdp_bitbang.c (hunk 3 above)
Comment (maxd-nordic) [reply drivers/dp/swdp_bitbang.c:612]: I can't spot the difference. What should I change?

**Review by tmon-nordic — DISMISSED** [review]

File: drivers/dp/swdp_bitbang.c (hunk 3 above)
Comment (jfischer-no) [reply drivers/dp/swdp_bitbang.c:613]: Okay:

```c
		if (ret) {
			return ret;
		}

		if (foo->bar) {
			/* */
		}
	}

	return -42
```

Not okay:

```c
		if (ret) {
			return ret;
		}
		if (foo->bar) {
			/* */
		}
	}
	return -42
```

**Review by mniestroj — DISMISSED** [review]

File: drivers/dp/swdp_bitbang.c (hunk 3 above)
Comment (maxd-nordic) [reply drivers/dp/swdp_bitbang.c:613]: Ah, you would like some space between the blocks, gotcha.

**Review by mniestroj — DISMISSED** [review]

File: drivers/dp/swdp_bitbang.c (hunk 3 above)
Comment (jfischer-no) [reply drivers/dp/swdp_bitbang.c]: Still not addressed, please take a look at the changes in sw_port_on().

Comment (maxd-nordic) [reply drivers/dp/swdp_bitbang.c]: it it okay now? @jfischer-no

**Review by jfischer-no — APPROVED** [review]

**Review by mniestroj — APPROVED** [review]

**Pull request merged**